docs/developer_guidelines.rst

```diff
@@ -41,9 +40,92 @@ Blocks:
 .. _pyflakes: https://pypi.python.org/pypi/pyflakes
 .. _should be grouped: https://www.python.org/dev/peps/pep-0008/#imports
 
+Code guidelines
+---------------
+Some guidelines to keep in mind when coding for Blocks. Some of these are simply
+preferences, others stem from particular requirements we have e.g. in order to
+serialize training progress, support Python 2 and 3 simultaneously, etc.
+
+Validating function arguments
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+In general, be Pythonic and use `duck typing`_. That is, don't check the type of
```

Comment:
Okay, I'll nuance it a bit. It's mostly to discourage checks like `isinstance(var, six.integer_types)` when a float, NumPy scalar, etc. would work too. Same thing for checks like `isinstance(var, list)` when a tuple, dictionary view, or any other iterable would work, etc. Checks such as `isinstance(var, TheanoSharedVariable)` make sense of course.